geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/Z3Iterator.scala

```diff
@@ -0,0 +1,92 @@
+package org.locationtech.geomesa.curve
+
+import com.google.common.primitives.Longs
+import org.apache.accumulo.core.client.{BatchScanner, IteratorSetting}
+import org.apache.accumulo.core.data.{ByteSequence, Key, Range => AccRange, Value}
+import org.apache.accumulo.core.iterators.{IteratorEnvironment, SortedKeyValueIterator}
+import org.apache.hadoop.io.Text
+
+class Z3Iterator extends SortedKeyValueIterator[Key, Value] {
+  var source: SortedKeyValueIterator[Key, Value] = null
+  var zmin: Long = -1L
+  var zmax: Long = -1L
+
+  var zlatmin: Long = -1L
+  var zlonmin: Long = -1L
+  var ztmin: Long = -1L
+
+  var zlatmax: Long = -1L
+  var zlonmax: Long = -1L
+  var ztmax: Long = -1L
+
+  var topKey: Key = null
+  var topValue: Value = null
+  val row = new Text()
+  val rowZBytes = Array.ofDim[Byte](8)
+
+  override def deepCopy(env: IteratorEnvironment): SortedKeyValueIterator[Key, Value] =
```

Comment:
we'll probably need to implement deep copy if we want to stack other iterators on top of this one. There's an example in org.locationtech.geomesa.core.iterators.RowSkippingIterator#deepCopy if you need one.